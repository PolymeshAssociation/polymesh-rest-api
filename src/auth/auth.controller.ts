import { Body, Controller, Post } from '@nestjs/common';
import { ApiNoContentResponse, ApiOkResponse, ApiOperation, ApiTags } from '@nestjs/swagger';

import { AuthService } from '~/auth/auth.service';
import { CreateApiKeyDto } from '~/auth/dto/create-api-key.dto';
import { DeleteApiKeyDto } from '~/auth/dto/delete-api-key.dto';
import { ApiKeyModel } from '~/auth/models/api-key.model';

@ApiTags('auth')
@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @ApiOperation({
    summary: 'Create API Key',
    description: 'This endpoint will create an API Key',
  })
  @ApiOkResponse({
    description: 'Details of the API key created',
    type: ApiKeyModel,
  })
  @Post('api-key/create')
  public async createApiKey(@Body() params: CreateApiKeyDto): Promise<ApiKeyModel> {
    return this.authService.createApiKey(params);
  }

  @ApiOperation({
    summary: 'Delete an API Key',
    description: 'This endpoint invalidates the given API key',
  })
  @ApiNoContentResponse({
    description: 'The API key is no longer valid',
  })
  @Post('/api-key/delete')
  public async deleteApiKey(@Body() params: DeleteApiKeyDto): Promise<void> {
    await this.authService.deleteApiKey(params);
  }
}
