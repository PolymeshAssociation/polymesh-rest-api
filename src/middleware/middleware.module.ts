/* istanbul ignore file */

import { DynamicModule, forwardRef, Module } from '@nestjs/common';

import { ConfidentialAccountsModule } from '~/confidential-accounts/confidential-accounts.module';
import { ConfidentialAssetsModule } from '~/confidential-assets/confidential-assets.module';
import { ConfidentialAssetsMiddlewareController } from '~/middleware/confidential-assets-middleware/confidential-assets-middleware.controller';

@Module({})
export class MiddlewareModule {
  static register(): DynamicModule {
    const controllers = [];

    const middlewareUrl = process.env.POLYMESH_MIDDLEWARE_V2_URL || '';

    if (middlewareUrl.length) {
      controllers.push(ConfidentialAssetsMiddlewareController);
    }

    return {
      module: MiddlewareModule,
      imports: [
        forwardRef(() => ConfidentialAssetsModule),
        forwardRef(() => ConfidentialAccountsModule),
      ],
      controllers,
      providers: [],
      exports: [],
    };
  }
}
