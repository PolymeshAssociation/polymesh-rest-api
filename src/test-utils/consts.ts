import { createMock } from '@golevelup/ts-jest';
import { BigNumber } from '@polymeshassociation/polymesh-sdk';
import { Account , PayingAccountType, TransactionStatus } from '@polymeshassociation/polymesh-sdk/types';

import { UserModel } from '~/users/model/user.model';

const signer = 'alice';
const did = '0x01'.padEnd(66, '0');
const dryRun = false;

const user = new UserModel({
  id: '-1',
  name: 'TestUser',
});

const resource = {
  type: 'TestResource',
  id: '-1',
} as const;

export const testAccount = createMock<Account>({ address: 'address' });
export const txResult = {
  transactions: ['transaction'],
  details: {
    status: TransactionStatus.Succeeded,
    fees: {
      gas: new BigNumber(1),
      protocol: new BigNumber(1),
      total: new BigNumber(1),
    },
    supportsSubsidy: false,
    payingAccount: {
      address: did,
      balance: new BigNumber(1),
      type: PayingAccountType.Caller,
    },
  },
};

export const testValues = {
  signer,
  did,
  user,
  resource,
  testAccount,
  txResult,
  dryRun,
};
