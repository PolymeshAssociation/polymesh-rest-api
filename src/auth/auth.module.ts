/* istanbul ignore file */

import { Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { APP_GUARD } from '@nestjs/core';
import { IAuthGuard, PassportModule } from '@nestjs/passport';

import { AuthController } from '~/auth/auth.controller';
import { AuthService } from '~/auth/auth.service';
import { createAuthGuard } from '~/auth/auth.utils';
import { ApiKeyStrategy } from '~/auth/strategies/api-key.strategy';
import { OpenStrategy } from '~/auth/strategies/open.strategy';
import { DatastoreModule } from '~/datastore/datastore.module';
import { UsersModule } from '~/users/users.module';

/**
 * responsible for the REST API's authentication strategies
 *
 * @note authorization has not yet been implemented - all users have full access
 */
@Module({
  imports: [
    ConfigModule,
    DatastoreModule.registerAsync(),
    UsersModule,
    PassportModule.register({
      session: false,
    }),
  ],
  providers: [
    AuthService,
    ApiKeyStrategy,
    OpenStrategy,
    {
      provide: APP_GUARD, // registers a global guard
      useFactory: (config: ConfigService): IAuthGuard => {
        const configuredStrategies = config.getOrThrow('AUTH_STRATEGY');
        return createAuthGuard(configuredStrategies);
      },
      inject: [ConfigService],
    },
  ],
  controllers: [AuthController],
  exports: [AuthService, PassportModule],
})
export class AuthModule {}
