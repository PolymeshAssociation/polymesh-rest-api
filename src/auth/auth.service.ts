import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { parseApiKeysConfig } from '~/auth/auth.utils';
import { CreateApiKeyDto } from '~/auth/dto/create-api-key.dto';
import { DeleteApiKeyDto } from '~/auth/dto/delete-api-key.dto';
import { ApiKeyModel } from '~/auth/models/api-key.model';
import { generateBase64Secret } from '~/common/utils';
import { User, UserId } from '~/users/user.consts';

@Injectable()
export class AuthService {
  private userStore: Record<UserId, User> = {};

  constructor(config: ConfigService) {
    const givenApiKeys = config.getOrThrow<string>('API_KEYS');
    const apiKeys = parseApiKeysConfig(givenApiKeys);

    const configuredUserId = 'configured-user';
    this.userStore[configuredUserId] = { id: configuredUserId, apiKeys };
  }

  public async createApiKey({ userId }: CreateApiKeyDto): Promise<ApiKeyModel> {
    const secret = await generateBase64Secret(32);

    if (!this.userStore[userId]) {
      this.userStore[userId] = { id: userId, apiKeys: [] };
    }
    this.userStore[userId].apiKeys.push(secret);

    return { userId, secret };
  }

  /**
   * Returns the associated user by an API key
   */
  public validateApiKey(apiKey: string): User | undefined {
    return this.getUserByApiKey(apiKey);
  }

  /**
   * @param apiKey - key to delete
   */
  public async deleteApiKey({ apiKey }: DeleteApiKeyDto): Promise<void> {
    const user = this.getUserByApiKey(apiKey);
    if (user) {
      user.apiKeys = user.apiKeys.filter(key => key !== apiKey);
    }
  }

  /**
   * @note, will likely need to be a abstracted with a "repository" type interface when a data store is added
   */
  private getUserByApiKey(apiKey: string): User | undefined {
    return Object.values(this.userStore).find(({ apiKeys }) => apiKeys.includes(apiKey));
  }
}
