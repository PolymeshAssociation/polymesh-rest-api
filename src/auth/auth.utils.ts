import { AuthGuard, IAuthGuard } from '@nestjs/passport';

import { AuthStrategy, authStrategyValues } from '~/auth/strategies/strategies.consts';

/**
 *  Creates an AuthGuard using the configured strategies
 */
export const createAuthGuard = (rawStrategy: string): IAuthGuard => {
  const strategies = parseAuthStrategyConfig(rawStrategy);
  return new (class extends AuthGuard(strategies) {})();
};

/**
 * transforms a raw auth strategy config into valid strategy values
 *
 * @throws if given invalid values
 */
const parseAuthStrategyConfig = (rawStrategyConfig: string): AuthStrategy[] => {
  const givenStrategies = rawStrategyConfig.split(',').map(strategy => strategy.trim());

  const filteredStrategies = givenStrategies.filter(isStrategyKey);

  if (filteredStrategies.length !== givenStrategies.length) {
    throw new Error(
      `Auth config error! "${givenStrategies}" contains an unrecognized option. Valid values are: "${authStrategyValues}"`
    );
  }

  return filteredStrategies.sort(cmpAuthStrategyOrder);
};

export const parseApiKeysConfig = (rawApiKeyConfig: string): string[] => {
  if (rawApiKeyConfig.trim() === '') {
    return [];
  }

  return rawApiKeyConfig.split(',').map(rawKey => rawKey.trim());
};

const isStrategyKey = (key: string): key is AuthStrategy => {
  return authStrategyValues.includes(key as AuthStrategy);
};

/**
 * A helper to be passed to `.sort`. The order is in which they will be evaluated.
 *
 * For example the the "open" strategy should always be last to allow for testing of other strategies
 */
const cmpAuthStrategyOrder = (a: AuthStrategy, b: AuthStrategy): number =>
  authStrategyValues.indexOf(a) - authStrategyValues.indexOf(b);
