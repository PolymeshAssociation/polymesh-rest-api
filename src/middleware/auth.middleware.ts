import { Injectable, NestMiddleware, UnauthorizedException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import passport from 'passport';

import { parseAuthStrategyConfig } from '~/auth/auth.utils';
import { User } from '~/users/user.consts';

@Injectable()
export class AuthMiddleware implements NestMiddleware {
  private readonly strategies: string[];

  constructor(configService: ConfigService) {
    const rawAuth = configService.getOrThrow<string>('AUTH_STRATEGY');
    this.strategies = parseAuthStrategyConfig(rawAuth);
  }

  use(req: unknown, res: unknown, next: () => void): void {
    passport.authenticate(this.strategies, { session: false }, (request, user: User) => {
      if (user) {
        next();
      } else {
        throw new UnauthorizedException('REST API user was not found');
      }
    })(req, res, next);
  }
}
