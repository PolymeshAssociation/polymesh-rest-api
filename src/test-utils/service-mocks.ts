/* istanbul ignore file */

import { createMock } from '@golevelup/ts-jest';
import { ValueProvider } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { AuthService } from '~/auth/auth.service';
import { ClaimsService } from '~/claims/claims.service';
import { ComplianceRequirementsService } from '~/compliance/compliance-requirements.service';
import { TrustedClaimIssuersService } from '~/compliance/trusted-claim-issuers.service';
import { ConfidentialAccountsService } from '~/confidential-accounts/confidential-accounts.service';
import { ConfidentialAssetsService } from '~/confidential-assets/confidential-assets.service';
import { ConfidentialTransactionsService } from '~/confidential-transactions/confidential-transactions.service';
import { DeveloperTestingService } from '~/developer-testing/developer-testing.service';
import { MetadataService } from '~/metadata/metadata.service';
import { NetworkService } from '~/network/network.service';
import { NftsService } from '~/nfts/nfts.service';
import { ProofServerService } from '~/proof-server/proof-server.service';
import { SubsidyService } from '~/subsidy/subsidy.service';
import { ServiceProvider } from '~/test-utils/types';
import { TransactionsService } from '~/transactions/transactions.service';
import { UsersService } from '~/users/users.service';

export class MockAssetService {
  findOne = jest.fn();
  findFungible = jest.fn();
  findNftCollection = jest.fn();
  findHolders = jest.fn();
  findDocuments = jest.fn();
  setDocuments = jest.fn();
  findAllByOwner = jest.fn();
  registerTicker = jest.fn();
  createAsset = jest.fn();
  issue = jest.fn();
  transferOwnership = jest.fn();
  redeem = jest.fn();
  freeze = jest.fn();
  unfreeze = jest.fn();
  controllerTransfer = jest.fn();
  getOperationHistory = jest.fn();
}

export class MockTransactionsService {
  submit = jest.fn();
  getSigningAccount = jest.fn();
}

export const mockTransactionsProvider = {
  provide: TransactionsService,
  useValue: new MockTransactionsService(),
};

export class MockComplianceRequirementsService {
  setRequirements = jest.fn();
  findComplianceRequirements = jest.fn();
  pauseRequirements = jest.fn();
  unpauseRequirements = jest.fn();
  deleteAll = jest.fn();
  deleteOne = jest.fn();
  add = jest.fn();
  modify = jest.fn();
}

export const mockComplianceRequirementsServiceProvider: ValueProvider<ComplianceRequirementsService> =
  {
    provide: ComplianceRequirementsService,
    useValue: createMock<ComplianceRequirementsService>(),
  };

export const mockDeveloperServiceProvider: ValueProvider<DeveloperTestingService> = {
  provide: DeveloperTestingService,
  useValue: createMock<DeveloperTestingService>(),
};

export class MockSigningService {
  public getAddressByHandle = jest.fn();
}

export class MockTickerReservationsService {
  findOne = jest.fn();
  reserve = jest.fn();
  transferOwnership = jest.fn();
  extend = jest.fn();
  findAllByOwner = jest.fn();
}

export class MockAuthorizationsService {
  findPendingByDid = jest.fn();
  findIssuedByDid = jest.fn();
  findOneByDid = jest.fn();
  accept = jest.fn();
  remove = jest.fn();
}

export class MockAccountsService {
  getAccountBalance = jest.fn();
  transferPolyx = jest.fn();
  getTransactionHistory = jest.fn();
  getPermissions = jest.fn();
  findOne = jest.fn();
  freezeSecondaryAccounts = jest.fn();
  unfreezeSecondaryAccounts = jest.fn();
  modifyPermissions = jest.fn();
  revokePermissions = jest.fn();
  getTreasuryAccount = jest.fn();
}

export class MockEventsService {
  createEvent = jest.fn();
  findOne = jest.fn();
}

export class MockSubscriptionsService {
  findAll = jest.fn();
  findOne = jest.fn();
  createSubscription = jest.fn();
  updateSubscription = jest.fn();
  batchMarkAsDone = jest.fn();
  batchBumpNonce = jest.fn();
}

export class MockNotificationsService {
  findOne = jest.fn();
  createNotifications = jest.fn();
  updateNotification = jest.fn();
}

export class MockHttpService {
  post = jest.fn();
  request = jest.fn();
}

export class MockScheduleService {
  addTimeout = jest.fn();
  addInterval = jest.fn();
  deleteInterval = jest.fn();
}

export class MockIdentitiesService {
  findOne = jest.fn();
  findTrustingAssets = jest.fn();
  findHeldAssets = jest.fn();
  addSecondaryAccount = jest.fn();
  createMockCdd = jest.fn();
  registerDid = jest.fn();
}

export class MockSettlementsService {
  findInstruction = jest.fn();
  createInstruction = jest.fn();
  affirmInstruction = jest.fn();
  rejectInstruction = jest.fn();
  findVenueDetails = jest.fn();
  findAffirmations = jest.fn();
  createVenue = jest.fn();
  modifyVenue = jest.fn();
  canTransfer = jest.fn();
  findGroupedInstructionsByDid = jest.fn();
  findVenuesByOwner = jest.fn();
  withdrawAffirmation = jest.fn();
  rescheduleInstruction = jest.fn();
}

export class MockClaimsService {
  findIssuedByDid = jest.fn();
  findAssociatedByDid = jest.fn();
  findCddClaimsByDid = jest.fn();
}

export class MockPortfoliosService {
  moveAssets = jest.fn();
  findAllByOwner = jest.fn();
  createPortfolio = jest.fn();
  deletePortfolio = jest.fn();
  updatePortfolioName = jest.fn();
  getCustodiedPortfolios = jest.fn();
  getTransactions = jest.fn();
  findOne = jest.fn();
  createdAt = jest.fn();
  setCustodian = jest.fn();
  quitCustody = jest.fn();
}

export class MockOfferingsService {
  findInvestmentsByTicker = jest.fn();
  findAllByTicker = jest.fn();
}

export class MockCorporateActionsService {
  findDefaultConfigByTicker = jest.fn();
  updateDefaultConfigByTicker = jest.fn();
  findDistributionsByTicker = jest.fn();
  findDistribution = jest.fn();
  createDividendDistribution = jest.fn();
  remove = jest.fn();
  payDividends = jest.fn();
  claimDividends = jest.fn();
  linkDocuments = jest.fn();
  reclaimRemainingFunds = jest.fn();
  modifyCheckpoint = jest.fn();
}

export class MockCheckpointsService {
  findAllByTicker = jest.fn();
  findSchedulesByTicker = jest.fn();
  findScheduleById = jest.fn();
  createByTicker = jest.fn();
  createScheduleByTicker = jest.fn();
  getAssetBalance = jest.fn();
  getHolders = jest.fn();
  deleteScheduleByTicker = jest.fn();
  findOne = jest.fn();
}

export class MockAuthService {
  createApiKey = jest.fn();
  deleteApiKey = jest.fn();
  validateApiKey = jest.fn();
}

export class MockTrustedClaimIssuersService {
  find = jest.fn();
  set = jest.fn();
  add = jest.fn();
  remove = jest.fn();
}

export const mockAuthServiceProvider = {
  provide: AuthService,
  useValue: new MockAuthService(),
};

export const mockUserServiceProvider: ValueProvider<UsersService> = {
  provide: UsersService,
  useValue: createMock<UsersService>(),
};

export const mockTrustedClaimIssuersServiceProvider: ValueProvider<TrustedClaimIssuersService> = {
  provide: TrustedClaimIssuersService,
  useValue: createMock<TrustedClaimIssuersService>(),
};

export const mockMetadataServiceProvider: ValueProvider<MetadataService> = {
  provide: MetadataService,
  useValue: createMock<MetadataService>(),
};

export const mockSubsidyServiceProvider: ValueProvider<SubsidyService> = {
  provide: SubsidyService,
  useValue: createMock<SubsidyService>(),
};

export const mockNftsServiceProvider: ValueProvider<NftsService> = {
  provide: NftsService,
  useValue: createMock<NftsService>(),
};

/**
 * Given a set of key values to use as config, will wrap and return as a Nest "provider" for config
 */
export const makeMockConfigProvider = (config: Record<string, unknown>): ServiceProvider => {
  return {
    useValue: {
      get: (key: string): unknown => config[key],
      getOrThrow: (key: string): unknown => {
        const value = config[key];
        if (value) {
          return value;
        } else {
          throw new Error(`mock config error: "${key}" was not found`);
        }
      },
    },
    provide: ConfigService,
  };
};

export const mockNetworkServiceProvider: ValueProvider<NetworkService> = {
  provide: NetworkService,
  useValue: createMock<NetworkService>(),
};

export const mockClaimsServiceProvider: ValueProvider<ClaimsService> = {
  provide: ClaimsService,
  useValue: createMock<ClaimsService>(),
};

export const mockConfidentialAssetsServiceProvider: ValueProvider<ConfidentialAssetsService> = {
  provide: ConfidentialAssetsService,
  useValue: createMock<ConfidentialAssetsService>(),
};

export const mockConfidentialAccountsServiceProvider: ValueProvider<ConfidentialAccountsService> = {
  provide: ConfidentialAccountsService,
  useValue: createMock<ConfidentialAccountsService>(),
};

export const mockConfidentialTransactionsServiceProvider: ValueProvider<ConfidentialTransactionsService> =
  {
    provide: ConfidentialTransactionsService,
    useValue: createMock<ConfidentialTransactionsService>(),
  };

export const mockProofServerServiceProvider: ValueProvider<ProofServerService> = {
  provide: ProofServerService,
  useValue: createMock<ProofServerService>(),
};
