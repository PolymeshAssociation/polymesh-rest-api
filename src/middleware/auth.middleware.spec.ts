import { Test, TestingModule } from '@nestjs/testing';

import { ApiKeyStrategy } from '~/auth/strategies/api-key.strategy';
import { OpenStrategy } from '~/auth/strategies/open.strategy';
import { AuthMiddleware } from '~/middleware/auth.middleware';
import { makeMockConfigProvider, mockAuthServiceProvider } from '~/test-utils/service-mocks';

describe('AuthMiddleware', () => {
  let middleware: AuthMiddleware;
  let apiKeyStrategy: ApiKeyStrategy;
  let openStrategy: OpenStrategy;

  beforeEach(async () => {
    const mockConfigValues = {
      AUTH_STRATEGY: 'apiKey,open',
    } as Record<string, string>;

    const mockConfigProvider = makeMockConfigProvider(mockConfigValues);

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        mockConfigProvider,
        mockAuthServiceProvider,
        ApiKeyStrategy,
        OpenStrategy,
        AuthMiddleware,
      ],
    }).compile();

    middleware = module.get<AuthMiddleware>(AuthMiddleware);

    apiKeyStrategy = module.get<ApiKeyStrategy>(ApiKeyStrategy);
    openStrategy = module.get<OpenStrategy>(OpenStrategy);
  });

  it('should be defined', () => {
    expect(middleware).toBeDefined();
  });

  it('should call configured strategies', () => {
    const apiSpy = jest.spyOn(apiKeyStrategy, 'authenticate');
    const openSpy = jest.spyOn(openStrategy, 'authenticate');
    const nextMock = jest.fn();

    middleware.use({}, {}, nextMock);

    expect(apiSpy).toHaveBeenCalled();
    expect(openSpy).toHaveBeenCalled();
    expect(nextMock).toHaveBeenCalled();
    apiSpy.mockRestore();
    openSpy.mockRestore();
  });

  it('should throw an UnauthorizedException if no user is found', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const exposedMiddleware = middleware as any;
    exposedMiddleware.strategies = ['apiKey'];

    const nextMock = jest.fn();

    expect(() => middleware.use({}, {}, nextMock)).toThrowErrorMatchingSnapshot();
    expect(nextMock).not.toHaveBeenCalled();
  });
});
