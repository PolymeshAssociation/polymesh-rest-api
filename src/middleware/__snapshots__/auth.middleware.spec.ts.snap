// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthMiddleware should throw an UnauthorizedException if no user is found 1`] = `"REST API user was not found"`;
